src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs

```diff
@@ -1,442 +1,199 @@
-                        {
-                            IAsyncCollector<TMessage> raw = builder(context, valueBindingContext);
-                            ICollector<TMessage> obj = new SyncAsyncCollectorAdapter<TMessage>(raw);
-                            return new AsyncCollectorValueProvider<ICollector<TMessage>, TMessage>(obj, raw, invokeString);
-                        };
-                    }
-                    else
-                    {
-                        // Bind to ICollector<T>. 
-                        // This needs both a conversion from T to TMessage and an Sync/Async wrapper
-                        argumentBuilder = DynamicInvokeBuildICollectorArgument(elementType, converterManager, builder, invokeString);
-                    }
-                }
-            }
+            bool singleDispatch;
+            var argumentBinding = BindingFactoryHelpers.GetTriggerArgumentBinding(bindingStrategy, parameter, converterManager, out singleDispatch);
 
-            if (parameter.IsOut)
+            var parameterDescriptor = new ParameterDescriptor
             {
-                Type elementType = parameter.ParameterType.GetElementType();
-
-                if (elementType.IsArray)
-                {
-                    if (elementType == typeof(TMessage[]))
-                    {
-                        argumentBuilder = (context, valueBindingContext) =>
-                        {
-                            IAsyncCollector<TMessage> raw = builder(context, valueBindingContext);
-                            return new OutArrayValueProvider<TMessage>(raw, invokeString);
-                        };
-                    }
-                    else
-                    {
-                        // out TMessage[]
-                        var e2 = elementType.GetElementType();
-                        argumentBuilder = DynamicBuildOutArrayArgument(e2, converterManager, builder, invokeString);
-                    }
-                }
-                else
+                Name = parameter.Name,
+                DisplayHints = new ParameterDisplayHints
                 {
-                    // Single enqueue
-                    //    out TMessage
-                    if (elementType == typeof(TMessage))
-                    {
-                        argumentBuilder = (context, valueBindingContext) =>
-                        {
-                            IAsyncCollector<TMessage> raw = builder(context, valueBindingContext);
-                            return new OutValueProvider<TMessage>(raw, invokeString);
-                        };
-                    }
-                    else
-                    {
-                        // use JSon converter
-                        // out T
-                        argumentBuilder = DynamicInvokeBuildOutArgument(elementType, converterManager, builder, invokeString);
-                    }
+                    Description = singleDispatch ? "message" : "messages"
```

Comment:
This is a general base class. I don't see how "message" applies generally to all extension bindings?